src/macros.rs

```diff
@@ -21,70 +21,130 @@ macro_rules! algos {
 
     ($($mod:ident)::+<$inner:ident>) => {
         algos!(@algo brotli ["brotli"] BrotliDecoder BrotliEncoder<$inner> {
-            /// The `level` argument here is typically 0-11.
-            pub fn new(reader: $inner, level: u32) -> Self {
-                let mut params = brotli::enc::backward_references::BrotliEncoderParams::default();
-                params.quality = level as _;
-                Self::from_params(reader, params)
+            pub fn new(reader: $inner) -> Self {
+                let params = brotli::enc::backward_references::BrotliEncoderParams::default();
+                Self {
+                    inner: crate::$($mod::)+generic::Encoder::new(
+                        reader,
+                        crate::codec::BrotliEncoder::new(params),
+                    ),
+                }
             }
         } {
-            pub fn from_params(inner: $inner, params: brotli::enc::backward_references::BrotliEncoderParams) -> Self {
+            /// The `level` argument here is typically 0-11.
+            pub fn with_quality(reader: $inner, level: crate::Compression) -> Self {
+                let mut params = brotli::enc::backward_references::BrotliEncoderParams::default();
+                match level {
+                    crate::Compression::Fastest => params.quality = 0,
+                    crate::Compression::Best => params.quality = 11,
+                    crate::Compression::Default => (),
+                }
                 Self {
                     inner: crate::$($mod::)+generic::Encoder::new(
-                        inner,
+                        reader,
                         crate::codec::BrotliEncoder::new(params),
                     ),
                 }
             }
         });
 
         algos!(@algo bzip2 ["bzip2"] BzDecoder BzEncoder<$inner> {
-            pub fn new(inner: $inner, level: bzip2::Compression) -> Self {
+            pub fn new(inner: $inner) -> Self {
                 Self {
                     inner: crate::$($mod::)+generic::Encoder::new(
                         inner,
-                        crate::codec::BzEncoder::new(level, 30),
+                        crate::codec::BzEncoder::new(bzip2::Compression::Default, 0),
+                    ),
+                }
+            }
+        } {
+            pub fn with_quality(inner: $inner, level: crate::Compression) -> Self {
+                Self {
+                    inner: crate::$($mod::)+generic::Encoder::new(
+                        inner,
+                        crate::codec::BzEncoder::new(level.into(), 0),
```

Comment:
I'm assuming we don't want to change the `work_factor` here, even for `Best` compression quality?